salt/engines/slack.py

```diff
@@ -129,600 +131,621 @@ def __virtual__():
-        return (key, val,)  # matched this group, return the group
-    log.info("Slack user {} denied trying to run {}".format(user, command))
-    return ()
-
-
-def commandline_to_list(cmdline_str, trigger_string):
-    """
-    cmdline_str is the string of the command line
-    trigger_string is the trigger string, to be removed
-    """
-    cmdline = salt.utils.args.shlex_split(cmdline_str[len(trigger_string):])
-    # Remove slack url parsing
-    #  Translate target=<http://host.domain.net|host.domain.net>
-    #  to target=host.domain.net
-    cmdlist = []
-    for cmditem in cmdline:
-        pattern = r'(?P<begin>.*)(<.*\|)(?P<url>.*)(>)(?P<remainder>.*)'
-        mtch = re.match(pattern, cmditem)
-        if mtch:
-            origtext = mtch.group('begin') + mtch.group('url') + mtch.group('remainder')
-            cmdlist.append(origtext)
+            groups_gen = itertools.chain(self._groups_from_pillar(groups_pillar_name).items(), use_groups.items())
+        except AttributeError:
+            log.warn("Failed to get groups from {}: {}".format(groups_pillar_name, self._groups_from_pillar(groups_pillar_name)))
+            log.warn("or from config: {}".format(use_groups))
+            groups_gen = []
+        for name, config in groups_gen:
+            log.info("Trying to get {} and {} to be useful".format(name, config))
+            ret_groups.setdefault(name, {
+                "users": set(), "commands": set(), "aliases": dict(), "default_target": dict(), "targets": dict()
+            })
+            try:
+                ret_groups[name]['users'].update(set(config.get('users', [])))
+                ret_groups[name]['commands'].update(set(config.get('commands', [])))
+                ret_groups[name]['aliases'].update(config.get('aliases', {}))
+                ret_groups[name]['default_target'].update(config.get('default_target', {}))
+                ret_groups[name]['targets'].update(config.get('targets', {}))
+            except IndexError:
+                log.warn("Couldn't use group {}. Check that targets is a dict and not a list".format(name))
+
+        log.debug("Got the groups: {}".format(ret_groups))
+        return ret_groups
+
+    def _groups_from_pillar(self, pillar_name):
+        '''pillar_prefix is the pillar.get syntax for the pillar to be queried.
+        Group name is gotten via the equivalent of using
+        ``salt['pillar.get']('{}:{}'.format(pillar_prefix, group_name))``
+        in a jinja template.
+
+        returns a dictionary (unless the pillar is mis-formatted)
+        XXX: instead of using Caller, make the minion to use configurable so there could be some
+             restrictions placed on what pillars can be used.
+        '''
+        caller = salt.client.Caller()
+        pillar_groups = caller.cmd('pillar.get', pillar_name)
+        # pillar_groups = __salt__['pillar.get'](pillar_name, {})
+        log.info("Got pillar groups {} from pillar {}".format(pillar_groups, pillar_name))
```

Comment:
Single-quotes here and on the next two lines please. :]

Reply:
Both fixed.  These were all introduced in a previous commit and I believe I've caught them all in this last push.